### PR TITLE
kernel/sched: Optimize handling for suspend(_current)

### PR DESCRIPTION
**Pull request by andyross**
k_thread_suspend() is a general call intended to halt any thread running in any state on any CPU, it's not particular lean.  And it's not really intended for use a general purpose application API, as asynchronous halt/stop/kill APIs are inherently race prone (e.g. you can suspend a thread holding a resource, there's no locking around the "decide to suspend the thread" that can be atomically released with the suspend, etc...  Semaphores and condition variables exist for a reason folks!).

But nonetheless, there's a notable benchmark in the tree now that uses exactly this API to mean "stop the current thread", and as it's a competitive thing we don't want to have to be "fixing" it for Zephyr.

So do a tiny, targetting bit of benchmark chasing: detect the case where we're suspending _current out of thread code on a non-SMP system and implement the minimal set of operations as an early exit.  This isn't too big, has no impact on SMP builds, and should make us look better to the cook kids at school.

### CODE REVIEW

**Comment by andyross** [comment]
As suggested in #81684, this is a shorter and more targetted attempt at "optimizing" a case that IMHO we shouldn't be optimizing at all.  In theory[1] k_thread_suspend(), for the specific case exercised in the thread_metric preemptive benchmark, can't possibly get any faster than this as this represents the minimal amount of work for the operation.

And it's short and doesn't impact kernel design anywhere else, allowing us to evolve design of things like sleep/suspend slowly and along cleaner aesthetic paths (and not out of benchmark chasing).

My twister run isn't done yet but AFAICT this hasn't broken anything.

[1] I don't have a proper rig to measure it and I ain't gonna.

**Comment by andyross** [comment]
(I do see a bunch of typos in the commit message and comment, but will wait for the CI results before re-pushing)

**Comment by andyross** [comment]
So, I know I said I wasn't going to run real tests, but I couldn't resist.  The test integration is super clean and obvious, and I have a frdm_k64f on my desk anyway.

Without this patch I get:

```
*** Booting Zephyr OS build v4.0.0-604-gf235ddb2cf26 ***
**** Thread-Metric Preemptive Scheduling Test **** Relative Time: 30
Time Period Total:  6692286

**** Thread-Metric Preemptive Scheduling Test **** Relative Time: 60
Time Period Total:  6691546

**** Thread-Metric Preemptive Scheduling Test **** Relative Time: 90
Time Period Total:  6691402

**** Thread-Metric Preemptive Scheduling Test **** Relative Time: 120
Time Period Total:  6691545
```

And with it:

```
** Booting Zephyr OS build v4.0.0-605-g81199ac320e4 ***
**** Thread-Metric Preemptive Scheduling Test **** Relative Time: 30
Time Period Total:  8257905

**** Thread-Metric Preemptive Scheduling Test **** Relative Time: 60
Time Period Total:  8256990

**** Thread-Metric Preemptive Scheduling Test **** Relative Time: 90
Time Period Total:  8256811

**** Thread-Metric Preemptive Scheduling Test **** Relative Time: 120
Time Period Total:  8256989
```

So about 23% faster context switch.  Which IMHO is not so shabby for a 7-line patch with minimal impact.

FWIW: the result reporting of this benchmark is as weird as its API choice.  The number out is the number of round trips through a "stack" of threads at different priorities.  Where each one wakes up the one higher until the top of the stack, and then they all go to sleep on the way back down.  So since there are five threads, that's four context switches to get to the top thread and four down, so each round trip is **eight** context switches.

And the number reported is the number of "eight-switch-round-trips" per time interval, which appears to be 30 seconds.  So `N/3.75` gives (modulo slop from measurement and rig execution) "context switches per second", or `3.75e9/N` for "ns per context switch".

**Comment by andyross** [comment]
Tests passed, so repush with a more recent rebase and typo fixing.

**Review by henrikbrixandersen — DISMISSED** [review]

**Review by peter-mitsis — COMMENTED** [review]
I am not opposed to the basic premise of this commit, which is to have an optimized path for k_thread_suspend(_current). However, at present the comments both in the commit messsage and the code strongly suggest that we are merely symptom chasing and trying to look good rather than improve and _be_ good. These benchmarks that are in the wild can help provide the necessary impetus for us to take a closer look at our code to trim, tighten and firm things up.

**Comment by andyross** [comment]
Yeah, I kinda knew things were going to go off the rails when I posted that rant.  Can we all take a deep breath?  It's just code.  I'm a nice guy and not trying to step on toes or call people names, I promise.

This gets back to the "good benchmarks or bad benchmarks" thing.  I'll just repeat that any number we get from thread_metric is "suspect", because it's not testing something that real world apps want to be fast.  It's just not.  And if we want to optimize for that (which we can, to pretty significant effect, c.f. this 7-line patch which shaves off 1/4 of the time we spent in "suspend-me"), we owe it to ourselves to be honest about what we're going and why.

Now, sure, I can rephrase that comment to say something neutral like "Known applications and benchmarks assume k_thread_suspend(_current); is the fastest way to suspend a thread, so handle that case specially".  But we both know that's not what we're actually doing, we're actually trying to get numbers from thread_metric to be higher, and none of us actually cares about the speed of k_thread_suspend() in the abstract because we know it's not a performance path.

Bottom line is that if we care about those thread_metric numbers, we should merge this patch, because it makes the "OS That Thread Metric Assumes" as fast as is practically possible.  If we don't care about thread_metric, we should find a benchmark we do care about and start writing to that.

File: kernel/sched.c [hunk 1]
```diff
@@ -499,6 +499,22 @@ void z_impl_k_thread_suspend(k_tid_t thread)
 {
 	SYS_PORT_TRACING_OBJ_FUNC_ENTER(k_thread, suspend, thread);
 
+	/* There are benchmarks in the wild which use
+	 * k_thread_suspend(_current) to force a context switch, which
+	 * is not really an appropriate use of the API (suspend is
+	 * inherently unsynchronized and deadlocky!).  But make it
+	 * fast just for numbers.
+	 */
+	if (thread == _current && !arch_is_in_isr() && !IS_ENABLED(CONFIG_SMP)) {
```
[the remark this answers is not in the document]
Comment (andyross) [reply kernel/sched.c:508]: Honestly no?  That seems exactly backwards.  We can only do that by adding complexity (and likely cycles) to already-known-to-be-frightently-subtle code.  This is one specific case, it's not like we don't need the general code below for all the other bits of the state space.  If there are dead code paths, the compiler will see them anyway.

**Comment by andyross** [comment]
> Yeah, I kinda knew things were going to go off the rails when I posted that rant. Can we all take a deep breath? It's just code. I'm a nice guy and not trying to step on toes or call people names, I promise.

Though to not retreat too much: I **really**, **really** hate that benchmark, not least because it's let us to some very weird places chasing performance paths in this obscure[1] corner of the scheduler.

[1] And, to repeat yet again, very much known to be dangerous!  Thread abort is on its third rewrite currently.  It's really not something we should be mucking with at all, much less for a questionable benchmark.

**Comment by peter-mitsis** [comment]
I'm not a fan of that benchmark either. I have a hard time putting into words what exactly it is that rubs me the wrong the way about it--it just seems _off / not quite right_. It has more to do with its assumptions and what it claims to be benchmarking vs what it is actually benchmarking.

That being said, the choice of its using the suspend/resume APIs is not an unreasonable choice to use to gauge the performance of that category of context switches--even though neither of us expect suspend nor resume to be hot paths in the field. What it does have going for it is that it bypasses kernel object overhead (small as it may be) that would be associated with semaphores, mutexes ....

If we can clean up the so-called _rant_ in both the commit message and comment, I am happy to +1 it as the premise is a reasonable and straightforward thing to to. (I will still be continuing to push for various other minor cleanups and enhancements).

File: kernel/sched.c (hunk 1 above)
[the remark this answers is not in the document]
Comment (peter-mitsis) [reply kernel/sched.c:508]: No worries. I can create a PR at a later time. On my local branch, this extra complexity allows me to increase the numbers from 7050442 to 7152170 assuming that I have not messed anything up.  (I have other changes on my branch that I am evaluating.) Incidentally, my the ThreadX numbers are in the neighborhood of 7320000, so we are getting closer to parity.

**Comment by andyross** [comment]
Fair enough, "professional" version up.  I still point out that trying to explain it like that is just fooling ourselves and our users.  But at the same time anyone who wants my opinions knows where to find me, or can follow the link to this PR to see what I really thought.

File: kernel/sched.c (hunk 1 above)
[the remark this answers is not in the document]
Comment (andyross) [reply kernel/sched.c:505]: Um... you can't be seeing impact from that.  The **only** use of k_thread_suspend() in the benchmark is to suspend the current thread, so the benchmark will only hit the new code and none of the cleanup you propose to the other paths.  This isn't the first time we've seen some counter-intuitive numbers from thread_metric.  Is it possible that your hardware has a cranky icache or something that is sensitive to alignment/layout and producing noisy numbers between rebuilds?  Or maybe there's something throttling the clock or another unit contending with the bus?

**Review by teburd — APPROVED** [review]

**Comment by nashif** [comment]
The main point of running this benchmark is that it gives us a point of comparison with other RTOSes, not because it is a great benchmark. For sure it misses many things we try to do with our own inhouse benchmarks in `tests/benchmark` and only touches the surface in many areas, i.e. the syncronisation benchmark does not really do synchronization, it just tests semaphore give/take performance from same thread if I remember correctly.

Anyways, this is the world of benchmarks unfortunately and this is the first time Zephyr is put in the spotlight like this. While the benchmark itself does not cover the "fast" APIs and real world usecases, it can be argued that by taking a sample of APIs and testing them, it can give you an overall idea about everything else in the OS. If an OS performs on the least used APIs, then good chance all other APIs and real world application would follow the same path.

Think of the 0-60 mph car benchmark, not something you would do on a daily basis with your cat, but still give you an idea about the car in general :)

We have learned about a few things we should improve in zephyr (not directly in the kernel) based on the report and benhcmark topic and many good points are being made here that we should summerize and have in some official response once we have further investigated and addressed some of those issues.

**Review by henrikbrixandersen — APPROVED** [review]

**Review by cfriedt — APPROVED** [review]

**Review by peter-mitsis — APPROVED** [review]

**Review by TaiJuWu — APPROVED** [review]

**Review by kartben — APPROVED** [review]
